### PR TITLE
Add/phone number react component

### PR DESCRIPTION
**Pull request by umurkontaci**
This PR is part for #7453 and dependent on #7454. It adds a phone-number component that 
- Can detect the country from the phone number
- Can format for national, international and NANPA formats
- Allows user to select their own country
- Uses vector based flags

It uses `flag-icon-css` to provide vector based flags.
### Teaser

ykdzcapvqh
### Testing

There is one example in `/devdocs` where you can test it. 

Make sure to test around the edge cases as well such as when there are few numbers in the input, or plenty of them.

One of the known quirks is when the user deletes / inserts a number between the characters.

If a fix proves to be complex or relies on `phone-number` library, it might be easier to both of the PRs in and do the fix in a follow up PR. This shouldn't be a problem since none of them are being used anywhere but devdocs.
### Design Notes

There are some countries in our country list that do not have a flag in SVG. They shouldn't be too many in numbers, but we might need to find and add those flags to our repository manually or to upstream `flag-icon-css`.

When the country flag changes, there is a small delay as the flag changes require a download, I've did a hack by putting a loading indicator _behind_ the flag so when it is invisible, the spinner is shown automatically. The problem is some of the flags are not at 100% opacity and you can see the spinner lying behind. Not sure what the solution is.

I have opted to use a regular `<select/>` element instead of a custom element that acts like a selector. This should give us more confidence for cross-compatibility. 

The flag as it is does not look clickable as it lacks an down arrow. We probably should add an arrow, but I wasn't sure how it would look like, we might also need to get rid of the borders around the flags as well. 

/cc: @klimeryk @aidvu @wensco @breezyskies @Automattic/sdev-feed 

Test live: https://calypso.live/devdocs/design/form-fields?branch=add/phone-number-react-component

### CODE REVIEW

**Comment by breezyskies** [comment]
This should probably get more eyes from outside Spectrum, since it has the potential to replace an existing component / pattern. CC'ing @mtias for feedback.

File: assets/stylesheets/style.scss [hunk 1]
```diff
@@ -1,6 +1,9 @@
 // External Dependencies
 @import '../../node_modules/react-virtualized/styles';
 
+$flag-icon-css-path: 'images/flags' !default;
+@import '../../node_modules/flag-icon-css/sass/flag-icon';
```
Comment (mtias) [inline assets/stylesheets/style.scss]: I'm not crazy about this css dependency here. Have you thought of any alternatives?

File: client/components/phone-input/index.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+		this.setState( {
+			selectedCountry: countries[ countryCode ],
+			freezeSelection: true
+		} );
+	},
+
+	render() {
+		const flagClasses = classNames(
+			'phone-input__flag-icon',
+			'flag-icon',
```
Comment (mtias) [inline client/components/phone-input/index.js]: These last two classes should be throwing warnings with our linter.

File: client/components/phone-input/style.scss [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	pointer-events: none;
+	z-index: 1;
+}
+
+.phone-input__flag-container {
+	position: absolute;
+	top: 0;
+	bottom: 0;
+	padding: 0;
+	background-color: #eaeaea;
```
Comment (mtias) [inline client/components/phone-input/style.scss]: This should be using the gray color vars.

File: assets/stylesheets/style.scss (hunk 1 above)
Comment (mtias) [inline assets/stylesheets/style.scss]: For instance, it'd be nice if we could render the svgs inline.

**Comment by gwwar** [comment]
The flags are being used in a pretty tiny space. Have we considered using a simple sprite sheet instead? This should avoid needing a spinner.

I ❤️ the overall concept

**Comment by folletto** [comment]
> When the country flag changes, there is a small delay as the flag changes require a download

So yes: 

> The flags are being used in a pretty tiny space. Have we considered using a simple sprite sheet instead? This should avoid the loading headaches

+1

I checked examples for size. [This](https://github.com/lafeber/world-flags-sprite) is 87k in its @2x resolution (32px).

Checking the SVG library this PR is using, there are some SVGs that are **massive** in comparison. Seems to me one of the uncommon scenarios where SVG might not be a good choice.

---

Some feedback:

(1) Flag looks odd in its borders — the flag area has an extra border, rounded too, that makes it look odd (`phone-input__flag-container`).

screen shot 2016-08-29 at 11 47 09

It's probably better if the item is slightly smaller so it just sits inside the input control, with just a right border.

(2) Default status — the control shouldn't default to a specific country, or, if it does, should try to default by geolocation / IP / language / other locating technique. 

However: I acknowledge this might be overkill for now, so I'd say that at least it should start with one flag or a default one, and a text that is generic, like "Type the phone number with + prefix" (which should be default, and overridable with a prop).

(3) Chevron — I'm not sure if it makes sense to include a dropdown more explicitly, but if it does we can show the chevron in its small size (`12px`).

(4) If I switch country, then I start typing a different country prefix it "forces" the selected one. It should instead swap the flag.

(5) If a user first types a number and then changes the country, we should do a smart-prefixing (this might be a separate PR). So the whole logic should be:

* If nothing is in the input form, switching country adds the correct +00 code.
* If something is in, but without +00, it adds the +00 country code on top.
* If something is in, with the +00 prefix, it swaps the prefix with the new one.

Note: this assumes we normalize for +00. I think it would be a good idea, but I also know it's not an idea without consequences. So take this advice with the required contextual reframing.

Summing up all the points above in terms of UI:

screen shot 2016-08-29 at 12 22 37

File: assets/stylesheets/style.scss (hunk 1 above)
Comment (umurkontaci) [inline assets/stylesheets/style.scss]: If we stick with SVGs, we can create a React component that renders them.

**Comment by umurkontaci** [comment]
@gwwar @folletto thanks for pitching in!

> The flags are being used in a pretty tiny space. Have we considered using a simple sprite sheet instead? This should avoid the loading headaches

I thought about this, but didn't want to create different sprites for high DPIs and the ones with 1x resolution looked bad.

I'm not against the idea of using sprites, but might need some help in creating sprites.



> Default status — the control shouldn't default to a specific country, or, if it does, should try to default by geolocation / IP / language / other locating technique.

I think it should always default to something, and I planned it to default to the user's geolocation. Enforcing this simplifies the code quite a lot by avoid a `null` state.


> "Type the phone number with + prefix"

As long as Geolocation works or the user selects something on their own, they don't need to add the + prefix, they can use their local format. This message might suppress that feature.



> Chevron — I'm not sure if it makes sense to include a dropdown more explicitly, but if it does we can show the chevron in its small size (12px).

I think we should add a chevron, or something that indicates that the flag is clickable. 



> (4) If I switch country, then I start typing a different country prefix it "forces" the selected one. It should instead swap the flag.

While I agree, sometimes the countries' prefixes overlap (like +35 and +359), or they can actually share the whole county code (NAFTA countries or Russia & Kazakhstan). 


> (5) If a user first types a number and then changes the country, we should do a smart-prefixing (this might be a separate PR).

Since we do support phone numbers in local formats and geolocation to guess the user's country, this won't be necessary :)


PS: I like the new border and chevron.

**Comment by gwwar** [comment]
> I thought about this, but didn't want to create different sprites for high DPIs and the ones with 1x resolution looked bad.

If there aren't any suitable flag sprite sheets available, I suppose we could script out something to convert the SVG flags into images using imagemagick or inkscape on the command line. After we have those available in a folder, we can use a tool like [glue](https://glue.readthedocs.io/en/latest/quickstart.html) to create the sprite sheet for us.

If we need to maintain the flag images and build the sprite sheet, I'd recommend adding it as a separate repo and then pulling in the dependency.

Sprite sheets still bring in a decent amount of css, so cc @mtias for whether it's worth it.

**Comment by umurkontaci** [comment]
Oh I didn't think about the CSS addition at all. If we don't mind having bunch of SVGs in our repository, SVGs might be a better solution. After all, most users won't be changing their country selection.

**Comment by mtias** [comment]
As it stands, the svgs are probably better (if they are rendered inline just for this component) since CSS is currently loaded everywhere. A bunch of separate requests for every svg doesn't sound great, though.

**Comment by mtias** [comment]
If we get AsyncLoad merged soon, the whole dropdown could be loaded on demand when needed.

**Comment by gwwar** [comment]
If we do render these inline, we'll probably need designer help simplifying down the flags:
1. This keeps the size down
2. IE has a terrible SVG implementation, so you'll likely get crazy glitches in Edge/IE11

**Comment by umurkontaci** [comment]
New Preview:
image
If the country flag is missing:
image

I've opted to use inline SVGs, some of which are quite big. I think the simplest solution would be to use rasterized images for huge SVGs if we can't compress them. 

I would like get this into the hands of our users this week under an A/B test to gather some performance/usage data and keep developing in parallel. Unless anyone have objections, I'm going to go with only inline SVGs option and implement a better solution before I finish the rollout.

File: .gitattributes [hunk 4]
```diff
@@ -0,0 +1 @@
+npm-shrinkwrap.json binary
```
Comment (umurkontaci) [inline .gitattributes]: This will prevent git from attempting to merge `npm-shrinkwrap.json` as the only way to fix this file is to run `make shrinkwrap`.

File: client/components/forms/docs/example.jsx [hunk 5]
```diff
@@ -188,7 +189,7 @@ var FormFields = React.createClass( {
 					</FormButtonsBar>
 				</Card>
 
-				<Card>
+				<Card id="form-section">
```
Comment (klimeryk) [inline client/components/forms/docs/example.jsx]: Any particular reason why this `id` was added? I don't see it reference anywhere in our code/assets. Maybe it's a leftover from a merge?

File: client/components/phone-input/country-flag.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import React from 'react';
```
Comment (klimeryk) [inline client/components/phone-input/country-flag.js]: Missing the `Internal/External dependencies` comments 😊 (that's like serious business, OK? 😀)

**Comment by klimeryk** [comment]
I must admit that I found the overall experience a bit confusing - maybe because I was testing it with a Polish phone number. But the way the numbers jump around when you input them is a bit weird. I was expecting something more along the lines "type whatever you want and after a delay/blur, we'll format it as we see fit" - so maybe that's why I didn't find this "format immediately, even if I obviously did not finish typing". Maybe a delay or waiting till the full number if input would improve this?

Poland has also proven to be a good test case, because it showed that the background behind the flag has to be something else than white/transparent - otherwise the flag itself is hard to read:
screen shot 2016-10-31 at 20 42 22

And last but not least - I've input a sample Polish phone number, standard 9 digit mobile number:
screen shot 2016-10-31 at 20 50 00
The way it's been formatted seems _very_ wrong to me - I don't know any Polish person that would format their mobile number like this. It should be a simple `123 456 789` or `123 45 67 89`. The way it's been formatted by the field would make sense for a landline (the first two numbers being the area code), but I think most of our users input their mobile numbers here. I wonder if other countries' formats are also mainly for landlines? Or if Poland is special in some way that landlines have different formats than mobile numbers?

File: client/components/phone-input/index.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+			freezeSelection: true
+		} );
+	},
+
+	render() {
+		return (
+			<div className="phone-input">
+				<input
+					placeholder={ formatNumber(
+						( this.state.selectedCountry.nationalPrefix || '' ) + '9876543210', this.state.selectedCountry
```
Comment (klimeryk) [inline client/components/phone-input/index.js]: When I select `Kosovo` from the list, I always get an error originating on this line:
`Uncaught TypeError: Cannot read property 'nationalPrefix' of undefined`. Not sure if this is the only country this is affected by (I should buy some lottery tickets if that's the case XD).

**Comment by umurkontaci** [comment]
> I think most of our users input their mobile numbers here

I don't think that's a valid assumption. Though I've updated the algorithm to pick a better suited formatting pattern now. So, it should format mobile and landline numbers accordingly.

**Comment by umurkontaci** [comment]
> But the way the numbers jump around when you input them is a bit weird. I was expecting something more along the lines "type whatever you want and after a delay/blur, we'll format it as we see fit"

I don't think formatting after helps the user actually. The purpose of the input is to help them type their number in a more confident / easy way. Formatting it afterwards doesn't help much. You're right on the fact that it makes it a bit confusing when the format changes in the middle, but with the recently algorithm update, that should be minimized.

**Comment by umurkontaci** [comment]
> I wonder if other countries' formats are also mainly for landlines? Or if Poland is special in some way that landlines have different formats than mobile numbers?

When it comes to phone numbers, every country does everything differently 😃

File: client/components/phone-input/index.js (hunk 7 above)
Comment (umurkontaci) [inline client/components/phone-input/index.js]: Almost :) Kosovo is in a disputed region and it was not recognized by Google. There two other entries (Canary Islands and Netherlands Antilles). As far as I know, Netherlands Antilles dissolved in 2010 and Canary Islands is an autonomous community and not an official country. 

We should update our country database at some point, but I've updated the code to be resilient against such discrepancies.

**Comment by klimeryk** [comment]
Thanks - the formatting is indeed much better now, mobile and landline numbers are formatted perfectly for Poland now 👍  

My [comment](https://github.com/Automattic/wp-calypso/pull/7456#issuecomment-257407111) about flags melding with the background still holds, though.

After some more testing, I find the field "too helpful" at times, especially when editing a number - the cursor jumps around, the numbers are moved before/after a dash, etc. I admit that I'm very biased, because personally I hate those kind of fields that try very hard to "correct" you, but end up getting in your way, instead of helping. If it's working correctly, it's a great thing, but with something as complex and diverse as phone numbers (thankfully not _as_ complex as addresses...) I think it's very hard to get it right.
Anyway, here are some puzzling behaviours I've noticed:

Trying to type out the placeholder - notice how the opening bracket I've typed confusingly disappears (and doesn't let me retype it):
mo6hxl4xlf

Editing a phone number - first I delete `4`. Notice how `3` moves into its place - kinda confusing, IMHO (even if I understand that's how the algorithm works). And when I type the new number, `2`, the cursor is sent to the end of the number 😕 
dzp1o23mpo

Removing the first digit of phone number, then trying to remove the country code (because why not? :D) - there's some weird stuff going on, suddenly a `4` appears, and the cursor jumps to the end of the number:
b60dznrbs1

All in all, it seems to me like it improves some cases, while making others more confusing and hard to perform. Personally, I was expecting a different approach - allow the user to input whatever they want, in whatever format they want (use dots to separate the digits? Sure thing, go ahead, I know what you mean!) and we'd have a robust validation and parsing logic backing this field that would try to "extract" the number from the user's input (for example, on blur or when submitting the form). If it fails or there's some uncertainty, we'd request the user to correct it. But that's just my uneducated opinion - I have no facts or research to back it up :) Knowing you, I'm sure you've done your homework and researched this and would love to hear your arguments for going with this approach 🙇 

Of course, for A/B testing purposes it's more than adequate - especially in comparison to the current approach. So my last paragraph there is just voicing my opinion on the whole approach and is in no way a blocker for testing this approach.

**Comment by klimeryk** [comment]
I'm adding the `Needs Design Review` label to get some input from real designers 😁

**Comment by umurkontaci** [comment]
@folletto Do you have any feedback on @klimeryk 's comment about flag blending in?

**Comment by folletto** [comment]
> Do you have any feedback on @klimeryk 's comment about flag blending in?

Let's ship this and review separately? ;)

We can consider adding a light border or light+sharp shadow, like `box-shadow: 0px 0px 1px 	lighten( $gray, 20% );`:


screen shot 2016-11-03 at 11 12 29

screen shot 2016-11-03 at 11 12 46

**Comment by drw158** [comment]
> As it stands, the svgs are probably better (if they are rendered inline just for this component) since CSS is currently loaded everywhere. A bunch of separate requests for every svg doesn't sound great, though.

Are SVG sprites an option?

**Comment by umurkontaci** [comment]
> Are SVG sprites an option?

SVG sprites are not different than regular sprites, comes with CSS bloat and it's going to be a 4MB download.

**Comment by rads** [comment]
Just a thought: it might be interesting to look at the distribution of existing phone numbers by country. You could provide a smaller SVG sprite sheet that contains the most common countries and the less common ones can be loaded as they currently are.

**Comment by umurkontaci** [comment]
I'm not sure what the sprites are helping to solve here.

**Comment by rads** [comment]
@umurkontaci: I was under the impression that we're trying to get rid of the loading animation without bulking up the initial page load. If not, you can ignore my previous comment. :smile:

**Comment by umurkontaci** [comment]
> I was under the impression that we're trying to get rid of the loading animation without bulking up the initial page load. If not, you can ignore my previous comment. 😄

Ah I see. I don't think it's going to cause us a lot of problems. We'll get the users' location from their geoIP, it's not very likely that they will change the country selection too often. I believe we're better off if we keep the flags separate.

**Comment by umurkontaci** [comment]
I've improved the algorithm to choose a more stable algorithm for NANPA countries and added a box-shadow.

**Comment by klimeryk** [comment]
Thanks! 👍 The flag is much more readable now with the box shadow, too 🙇 
OK, final pass at this. Just to be clear, none of these are blockers, IMHO. They can be fixed while you prepare the A/B test 😉 

* The placeholder often (for the less popular countries?) does not reflect how the phone number will actually get formatted. Maybe you can pass the placeholder number through the formatter before setting it? That way we'll have consistency :)
* I'm not sure if the placeholder is constant, but, for example, for Poland it has 10 digits, while none of numbers I know can have that many digits in Poland. Both mobile and landline ones have 9, so the placeholder can be confusing.
* You can _sometimes_ input non-numeric characters, it's hard to pin point what triggers this, looks like you can input any character if it's one of the first three?
rrbyruhngi

**Comment by umurkontaci** [comment]
> The placeholder often (for the less popular countries?) does not reflect how the phone number will actually get formatted. Maybe you can pass the placeholder number through the formatter before setting it? That way we'll have consistency :)

The placeholder numbers gets through the formatter actually, if you type the exact same thing, you'll get the same formatting (NANPA countries have different formatting when there's a leading "1" or not: leading "1" uses international formatting, but not using any leading country code uses national formatting.



> I'm not sure if the placeholder is constant, but, for example, for Poland it has 10 digits, while none of numbers I know can have that many digits in Poland. Both mobile and landline ones have 9, so the placeholder can be confusing.

Placeholder is constant at this time, we can look onto improving the placeholder in a later iteration.



> You can sometimes input non-numeric characters, it's hard to pin point what triggers this, looks like you can input any character if it's one of the first three?

This happens because I'm not formatting until there are 3 characters, this is a bug. Will definitely fix before the launch.





> Edit/Added: What's the proper way of getting back the value of the phone number? I see that the onChange callback gets the formatted number, but it's not guaranteed that that number will contain the country code. I imagined that the logic for that will be embedded somewhere in the component and the code that uses that component won't have to worry about those things?

There will be, I'm already working on this on another branch. (This was getting too big anyway) There'll be a way to extract the value in E.164 format and the field value as it looks like.


Thanks for the review!

**Comment by umurkontaci** [comment]
Thanks for everyone with their support on this!

**Comment by mtias** [comment]
@umurkontaci mind adding this one to devdocs?

**Comment by mtias** [comment]
Sorry, I see it's bundled with forms. I think we should start separating this into individual example files.
